debugger/gdb-sgx-plugin/thread.py

```diff
@@ -0,0 +1,326 @@
+
+        INSTANCE_DEBUG_EXT.add_thread(pid, tid, ppid, tid_target)
+        return False
+
+class ThreadCreateBreakpoint(gdb.Breakpoint):
+    def __init__(self):
+        """
+        Set a breakpoing @ kernel/thread.c::_run_thread, this should be right before myst_setjmp(&thread->jmpbuf)
+        Choosing this location because this is where the newly created thread about to run
+        We can get the new thread pid/tid here
+        """
+        super(ThreadCreateBreakpoint, self).__init__('kernel/thread.c:_run_thread', internal=False)
+
+    def stop(self):
+        # New thread info
+        ppid = int(gdb.parse_and_eval('((struct run_thread_arg*)arg_)->thread->process->ppid'))
+        pid = int(gdb.parse_and_eval('((struct run_thread_arg*)arg_)->thread->process->pid'))
+        tid = int(gdb.parse_and_eval('((struct run_thread_arg*)arg_)->thread->tid'))
+        tid_target = int(gdb.parse_and_eval('((struct run_thread_arg*)arg_)->target_tid'))
+        if DEBUG:
+            print("Created thread: ppid, pid, tid, tid_target", ppid, pid, tid, tid_target)
+        INSTANCE_DEBUG_EXT.add_thread(pid, tid, ppid, tid_target)
+
+        # Add _main_thread only once as it won't change often
+        if not INSTANCE_DEBUG_EXT.main_thread_tracked():
+            main_ppid = int(gdb.parse_and_eval('_main_thread->process->ppid'))
+            main_pid = int(gdb.parse_and_eval('_main_thread->process->pid'))
+            main_tid = int(gdb.parse_and_eval('_main_thread->tid'))
+            main_tid_target = int(gdb.parse_and_eval('_main_thread->target_tid'))
+            if DEBUG:
+                print("Main thread: ppid, pid, tid, tid_target", main_ppid, main_pid, main_tid, main_tid_target)
+            INSTANCE_DEBUG_EXT.add_thread(main_pid, main_tid, main_ppid, main_tid_target)
+
+        INSTANCE_DEBUG_EXT.copy_parent_process(pid, ppid)
+        return False
+
+class ThreadExitBreakpoint(gdb.Breakpoint):
+    def __init__(self):
+        """
+        Set a breakpoint @ kernel/thread.c::_run_thread, where myst_setjmp returns non-zero
+        Choosing this location because this is where the thread terminates
+
+        Breakpoint location reference
+
+        - kernel/thread.c:_run_thread
+        976 ...
+        977 /* Start time tracking for this thread */
+        978 myst_times_start();
+        979
+        980 /* Jump back here from exit */
+        981 if (myst_setjmp(&thread->jmpbuf) != 0)
+        982 {
+        983     /* ---------- running C-runtime thread descriptor ---------- */
+        984
+   b -> 985     assert(myst_gettid() != -1);
+        986
+        987     /* restore the target thread descriptor */
+        988     myst_set_fsbase(thread->target_td);
+        989
+        990     /* ---------- running target thread descriptor ---------- */
+        991 ...
+        """
+        super(ThreadExitBreakpoint, self).__init__('kernel/thread.c:985', internal=False)
```

Comment:
I would recommend adding debugger hooks here, instead of line number breakpoints. We use hooks for module load and unload too. See `oe_debug_notify_library_loaded` and `oe_debug_notify_library_unloaded` usage in tools/myst/host/host.c

Reply:
Created a hook `myst_debug_hook_thread_exit`